### ISSUE TITLE
AI invasion handling

### ISSUE THREAD

**Issue by Grummel7** [freeorion/freeorion#3745]
I'm sure many have wondered why the AI produces so many troop ships, flies them to random places and then seems to ignore them.
I've dug into it and the answer is complicated. #3735 was a part of it, but that was rather new. The other reasons go deeper.
The following multiplayer save file demonstrate a lot of them (check the Green AI): [GreenNotUsingTroopers.mps.gz](https://github.com/freeorion/freeorion/files/8256287/GreenNotUsingTroopers.mps.gz)

It was created before the fix of #3735. When you play it now with the current master, the AI will immediately send a lot of troop ships (much more than needed) to Sunce α I and successfully conquer it. Then it stop invading again. (btw, having systems called Sunce and Sonce next to each other can be confusing ;-)

Reasons for this behaviour:
1. When assigning troop fleets, `FleetUtilsAI.get_fleets_for_mission` is called with `target_stats = {"rating": 10, "troopCapacity": troops_needed + _TROOPS_SAFETY_MARGIN}`. Troop ships have no battle rating, so the target is never reached and the function assigns all available troop ships to Sunce α I and then finds that they at least fulfill the min_stats.
2. `AIFleetMission._check_retarget_invasion` does not check whether the mission target is still valid. After successful invasion extra troop ships remain with a mission to invade Sunce α I (but the don't fly there, since it now has shields, see below).
3. Some troop fleets are scheduled to invade the Acirema planet Alfirk α I. The planet has shields, but there is a ship keeping them down. The problem is that the ship's mission is MissionType.PROTECT_REGION, not MissionType.SECURE: `fleet_orders.py:634 : trooper_move_reqs_met() holding Invasion fleet 19467 before leaving supply because target (P_426<Alfirk α I>) has nonzero max shields and there is not yet a military fleet assigned to secure the target system.`

Point 1 could be fixed easily by removing the rating condition. It looks intentional, though. Does anyone have an idea?

Point 2: _check_retarget_invasion cries for refactoring. It has lots of returns and TODOs and only if no return is triggered, it does some retargeting. Also there is this:
```
        last_sys_target = INVALID_ID
        if orders:
            last_sys_target = orders[-1].target.id
        if last_sys_target == fleet.systemID:
            return  # TODO: check for best local target
```
The second if never triggers, because the last target of an invasion mission is a planet.

Point 3: This is the most puzzling one to me. Shouldn't the AI change the mission type of the ship when it decides to invade? Has it ever done so, and if so, why is it not doing it now? I guess I can add a special handling to change PROTECT_REGION into SECURE, but that there might be a bigger issue hidden behind it.

**Comment by Morlic-fo** [comment]
> Point 1 could be fixed easily by removing the rating condition. It looks intentional, though. Does anyone have an idea?

Probably existed at some point to add some escort or to shoot down some shields. I'd expect it isn't necessary.
I assume it went unnoticed because the AI would usually not have that many spare troop ships.



> Point 3: This is the most puzzling one to me. Shouldn't the AI change the mission type of the ship when it decides to invade? Has it ever done so, and if so, why is it not doing it now? I guess I can add a special handling to change PROTECT_REGION into SECURE, but that there might be a bigger issue hidden behind it.

It probably simply assigned the PROTECT_REGION mission much earlier because it didn't have any concrete invasion plans for the planet yet because it assumed it had insufficient troops. So it really just protects its colony in the system there.
I don't think the AI ever checked if it could repurpose existing fleets in the same system but would rather ask for new ships if it needs some... If the AI allocates a new fleet for an invasion target, I'd expect that fleet will get the SECURE mission.

I think it should be fine to repurpose such a ship.

**Comment by Grummel7** [comment]
Also there are still some problems with the Invasion handling, the discussed points have been fixed with #3754, so I think this Issue can be closed.
